indexprovider/wrapper.go

```diff
@@ -35,36 +36,48 @@ var shardRegMarker = ".boost-shard-registration-complete"
 var defaultDagStoreDir = "dagstore"
 
 type Wrapper struct {
-	cfg         lotus_config.DAGStoreConfig
-	enabled     bool
-	dealsDB     *db.DealsDB
-	legacyProv  lotus_storagemarket.StorageProvider
-	prov        provider.Interface
-	dagStore    *dagstore.Wrapper
-	meshCreator idxprov.MeshCreator
+	cfg            lotus_config.DAGStoreConfig
+	enabled        bool
+	dealsDB        *db.DealsDB
+	legacyProv     lotus_storagemarket.StorageProvider
+	prov           provider.Interface
+	dagStore       *dagstore.Wrapper
+	meshCreator    idxprov.MeshCreator
+	bitswapEnabled bool
 }
 
-func NewWrapper(cfg lotus_config.DAGStoreConfig) func(lc fx.Lifecycle, r repo.LockedRepo, dealsDB *db.DealsDB,
+func NewWrapper(cfg *config.Boost) func(lc fx.Lifecycle, r repo.LockedRepo, dealsDB *db.DealsDB,
 	legacyProv lotus_storagemarket.StorageProvider, prov provider.Interface, dagStore *dagstore.Wrapper,
 	meshCreator idxprov.MeshCreator) *Wrapper {
 
 	return func(lc fx.Lifecycle, r repo.LockedRepo, dealsDB *db.DealsDB,
 		legacyProv lotus_storagemarket.StorageProvider, prov provider.Interface, dagStore *dagstore.Wrapper,
 		meshCreator idxprov.MeshCreator) *Wrapper {
-		if cfg.RootDir == "" {
-			cfg.RootDir = filepath.Join(r.Path(), defaultDagStoreDir)
+		if cfg.DAGStore.RootDir == "" {
+			cfg.DAGStore.RootDir = filepath.Join(r.Path(), defaultDagStoreDir)
 		}
 
 		_, isDisabled := prov.(*DisabledIndexProvider)
-		return &Wrapper{
-			dealsDB:     dealsDB,
-			legacyProv:  legacyProv,
-			prov:        prov,
-			dagStore:    dagStore,
-			meshCreator: meshCreator,
-			cfg:         cfg,
-			enabled:     !isDisabled,
+		w := &Wrapper{
+			dealsDB:        dealsDB,
+			legacyProv:     legacyProv,
+			prov:           prov,
+			dagStore:       dagStore,
+			meshCreator:    meshCreator,
+			cfg:            cfg.DAGStore,
+			bitswapEnabled: cfg.Dealmaking.BitswapPeerID != "",
+			enabled:        !isDisabled,
 		}
+		// announce all deals on startup in case of a config change
+		lc.Append(fx.Hook{
+			OnStart: func(ctx context.Context) error {
+				go func() {
+					_ = w.IndexerAnnounceAllDeals(ctx)
```

Comment:
Looking at how this function is implemented it looks like it will announce
- all legacy deals that are in a sealing state (but not in an error or completed state)
- all new deals that have not already been announced (and are not in an error or completed state)

Is that what we want to happen here?

Also keep in mind that some providers have up to 100k deals.

Reply:
If there is no change to the protocols supported, this is essentially a no-op -- the call to announce will error with ErrAlreadyPublished.

per @willscott at the moment there is no more efficient way to do this.

Reply:
Ok but I don't think this is going to re-announce deals that are not in one of the two categories I mentioned. For example non-legacy deals that have already been announced.

Reply:
ok I now have more questions having read the code. I'm writing it up in a new issue. Basically, based on the behavior of the code as it exists, I really can't figure out what IndexerAnnounceAllDeals was intended to do.

The intent of the go-fil-markets code looks like a re-announce, while the code in boost looks like a first time announce, which will include several deals that really shouldn't be announced yet (i.e. haven't even had a data transfer)

Will document in more detail in a seperate issue.

Reply:
ok @dirkmc I think I've put this in a reasonable state. There are two new commits --
https://github.com/filecoin-project/go-fil-markets/pull/752/commits/5dd7bc10ed7d3e2e9b67f5dfa807a3e58ef876b2
- this removes StorageDealStaged from the list of states that announce, because StorageDealStaged is the stage where they get announces (vaguely equivalent to dealcheckpoints.AddedPiece in Boost)
- otherwise, I think the go-fil-markets logic is correct --
   - StorageDealActive describes the entire time the deal is on chain, which seems like a reasonable bound keep the announcement alive, and it is already among the states that get reannounced. The actual "completion states" are either errors or after the deal has actually expired (i.e. we're past SectorEndEpoch).
   
 https://github.com/filecoin-project/boost/pull/839/commits/63ba7018a5369e68d26bd15a3a0361aae5cbe614
 - this changes the condition for announce to essentially be only those deals that are at the IndexedAndAnnounced. Technically, it filters out all that are < IndexedAndAnnouced and >= Complete, which is currently the same as != IndexedAndAnnounced but I thought it'd be useful to treat this as a range in case we add something between IndexedAndAnnounced and Complete in the future.

Reply:
this addresses most of the issues in #882 though not all of them. It certainly shouldn't make things worse. :)